File: Safari Animals Image Recognition Deep Learning/ui/src/actions/Annotations.ts
import { failIfNotOk, StatusResponse } from "./StatusResponse";

export const Undetected = "undetected";

export interface Annotation {
  id: number;
  file_name: string;
  annotated_file_name: string | undefined;
  cropped_file_name: string | undefined;
  bbox: Array<number>;
  species_confidence: number;
  predicted_species: string;
  predicted_name: string;
  accepted: boolean | undefined;
  ignored: boolean | undefined;
}

export function compareAnnotations(a: Annotation, b: Annotation): number {
  if (a.predicted_species != b.predicted_species) {
    switch (true) {
      case a.predicted_species === Undetected:
        return 1;
      case b.predicted_species === Undetected:
        return -1;
      default:
        return a.predicted_species.localeCompare(b.predicted_species);
    }
  }
  switch (true) {
    case a.predicted_species === Undetected:
      return 1;
    case b.predicted_species === Undetected:
      return -1;
    default:
      return a.predicted_species.localeCompare(b.predicted_species);
  }
}

export function fetchAnnotations(
  collectionID: string
): Promise<Array<Annotation>> {
  interface AnnotationsResponse extends StatusResponse {
    annotations: Array<Annotation>;
  }

  return fetch(`/api/v1/annotations?collectionID=${collectionID}`, {
    method: "GET",
    headers: { Accept: "application/json" },
  })
    .then((resp) => resp.json())
    .then((data: AnnotationsResponse) => {
      failIfNotOk(data);
      return data.annotations;
    });
}

export function submitAnnotations(
  collectionID: string,
  annotations: Array<Annotation>
): Promise<void> {
  return fetch(`/api/v1/annotations?collectionID=${collectionID}`, {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json" },
    body: JSON.stringify(annotations),
  })
    .then((resp) => resp.json())
    .then((data: StatusResponse) => failIfNotOk(data));
}

File: Safari Animals Image Recognition Deep Learning/api/data_models/annotations.py
import json
from typing import TypedDict, Optional, List

from api.clients.redis_client import redis_client

REDIS_KEY = 'annotations'


class Annotation(TypedDict):
    id: str
    file_name: str
    annotated_file_name: Optional[str]
    cropped_file_name: Optional[str]
    bbox: Optional[List[float]]
    species_confidence: float
    predicted_species: str
    predicted_name: str
    accepted: bool
    ignored: bool


def truncate_annotations_for_collection(collection_id: str) -> None:
    key = __key_for_collection(collection_id)
    redis_client.delete(key)


def save_annotations_for_collection(collection_id: str, annotations: List[Annotation]) -> None:
    key = __key_for_collection(collection_id)
    for annotation in annotations:
        redis_client.hset(key, annotation['id'], json.dumps(annotation))


def read_annotations_for_collection(collection_id: str) -> List[Annotation]:
    key = __key_for_collection(collection_id)
    annotations = []
    for annotation_json in redis_client.hvals(key):
        annotations.append(json.loads(annotation_json))
    return annotations


def __key_for_collection(collection_id: str) -> str:
    return f'{REDIS_KEY}:collections:{collection_id}'
